crates/biome_js_analyze/tests/specs/nursery/useCollapsedIf/invalid.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Safe fixes:
+ */
+
+if (condition) {
+	if (anotherCondition) {
+		// ...
+	}
+}
+
+if (condition) {
+	if (anotherCondition) {
+		// ...
+	}
+} else {
+	// ...
+}
+
+if (condition) // Comment
+	if (anotherCondition)
+		doSomething();
+
+// Inner one is `JsBlockStatement`
+if (condition) if (anotherCondition) {
+	// ...
+}
+
+// Outer one is `JsBlockStatement`
+if (condition) {
+	if (anotherCondition) doSomething();
+}
+
+// No `JsBlockStatement`
+if (condition) if (anotherCondition) doSomething();
+
+// `JsEmptyStatement`
+if (condition) if (anotherCondition);
+
+// Nested
+if (a) {
+	if (b) {
+		// ...
+	}
+} else if (c) {
+	if (d) {
+		// ...
+	}
+}
+
+// Need parenthesis
+function* foo() {
+	if (a || b)
+		if (a ?? b)
+			if (a ? b : c)
+				if (a = b)
+					if (a += b)
+						if (a -= b)
+							if (a &&= b)
+								if (yield a)
+									if (a, b);
+}
+
+// Should not add parenthesis
+async function foo() {
+	if (a)
+		if (await a)
+			if (a.b)
+				if (a && b);
+}
+
+// Don't case parenthesis in outer test
+if (((a || b))) if (((c || d)));
+
+// Semicolon
+// TODO
+// if (a) {
+// 	if (b) foo()
+// }
+// [].forEach(bar)
```

Comment:
@togami2864
Thank you for your review! I added test cases from `unicorn/no-lonely-if` and fixed the rule to support cases that I missed on the first attempt.

Now this is an only failing case that I found. Because of ASI rule, we need to insert a semicolon before `[]`, after merging the inner `if` to the parent one:

```js
   if (a && b) foo()
   [].forEach(bar)
// ^ Syntax Error
```

Are there any utils or some related code in Biome to check semicolon is needed or insert it to the needed place?

Reply:
You can make new node with the token like this

```rs
   let new_node = AnyJsStatement::from(
        js_expression_statement(child_test)
            .with_semicolon_token(token(T![;]))
            .build(),
    );
```

There is no such utility, and it would be challenging to write one just for this edge case. I think you can skip this case or mark it as a to-do.

Reply:
Thank you 🙏 I already marked that case as to-do, so I will keep it as-is.